File: backend/app/db/postgres/models/resource_policies.py
from __future__ import annotations

import enum
import uuid

from sqlalchemy import (
    Boolean,
    Column,
    DateTime,
    Enum as SQLEnum,
    ForeignKey,
    Index,
    Integer,
    String,
    Text,
    UniqueConstraint,
    and_,
    text,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from ..base import Base


def _enum_values(enum_cls):
    return [item.value for item in enum_cls]


class ResourcePolicyPrincipalType(str, enum.Enum):
    ORGANIZATION_USER = "organization_user"
    PUBLISHED_APP_ACCOUNT = "published_app_account"
    EMBEDDED_EXTERNAL_USER = "embedded_external_user"


class ResourcePolicyResourceType(str, enum.Enum):
    AGENT = "agent"
    TOOL = "tool"
    KNOWLEDGE_STORE = "knowledge_store"
    MODEL = "model"


class ResourcePolicyRuleType(str, enum.Enum):
    ALLOW = "allow"
    QUOTA = "quota"


class ResourcePolicyQuotaUnit(str, enum.Enum):
    TOKENS = "tokens"


class ResourcePolicyQuotaWindow(str, enum.Enum):
    MONTHLY = "monthly"


class ResourcePolicySet(Base):
    __tablename__ = "resource_policy_sets"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    organization_id = Column(UUID(as_uuid=True), ForeignKey("organizations.id", ondelete="CASCADE"), nullable=False, index=True)
    name = Column(String, nullable=False)
    description = Column(Text, nullable=True)
    is_active = Column(Boolean, nullable=False, default=True, server_default=text("true"))
    created_by = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), nullable=True, index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    organization = relationship("Organization")
    creator = relationship("User")
    rules = relationship("ResourcePolicyRule", back_populates="policy_set", cascade="all, delete-orphan")

    __table_args__ = (
        UniqueConstraint("organization_id", "name", name="uq_resource_policy_sets_organization_name"),
    )


class ResourcePolicySetInclude(Base):
    __tablename__ = "resource_policy_set_includes"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    parent_policy_set_id = Column(
        UUID(as_uuid=True),
        ForeignKey("resource_policy_sets.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    included_policy_set_id = Column(
        UUID(as_uuid=True),
        ForeignKey("resource_policy_sets.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)

    parent_policy_set = relationship("ResourcePolicySet", foreign_keys=[parent_policy_set_id])
    included_policy_set = relationship("ResourcePolicySet", foreign_keys=[included_policy_set_id])

    __table_args__ = (
        UniqueConstraint(
            "parent_policy_set_id",
            "included_policy_set_id",
            name="uq_resource_policy_set_include_edge",
        ),
    )


class ResourcePolicyRule(Base):
    __tablename__ = "resource_policy_rules"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    policy_set_id = Column(
        UUID(as_uuid=True),
        ForeignKey("resource_policy_sets.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    resource_type = Column(
        SQLEnum(ResourcePolicyResourceType, values_callable=_enum_values),
        nullable=False,
        index=True,
    )
    resource_id = Column(String, nullable=False, index=True)
    rule_type = Column(
        SQLEnum(ResourcePolicyRuleType, values_callable=_enum_values),
        nullable=False,
        index=True,
    )
    quota_unit = Column(
        SQLEnum(ResourcePolicyQuotaUnit, values_callable=_enum_values),
        nullable=True,
    )
    quota_window = Column(
        SQLEnum(ResourcePolicyQuotaWindow, values_callable=_enum_values),
        nullable=True,
    )
    quota_limit = Column(Integer, nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    policy_set = relationship("ResourcePolicySet", back_populates="rules")

    __table_args__ = (
        Index(
            "uq_resource_policy_rules_allow_resource",
            "policy_set_id",
            "resource_type",
            "resource_id",
            unique=True,
            postgresql_where=and_(rule_type == ResourcePolicyRuleType.ALLOW),
            sqlite_where=and_(rule_type == ResourcePolicyRuleType.ALLOW),
        ),
        Index(
            "uq_resource_policy_rules_quota_resource",
            "policy_set_id",
            "resource_type",
            "resource_id",
            "quota_unit",
            "quota_window",
            unique=True,
            postgresql_where=and_(rule_type == ResourcePolicyRuleType.QUOTA),
            sqlite_where=and_(rule_type == ResourcePolicyRuleType.QUOTA),
        ),
    )


class ResourcePolicyAssignment(Base):
    __tablename__ = "resource_policy_assignments"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    organization_id = Column(UUID(as_uuid=True), ForeignKey("organizations.id", ondelete="CASCADE"), nullable=False, index=True)
    principal_type = Column(
        SQLEnum(ResourcePolicyPrincipalType, values_callable=_enum_values),
        nullable=False,
        index=True,
    )
    policy_set_id = Column(
        UUID(as_uuid=True),
        ForeignKey("resource_policy_sets.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=True, index=True)
    published_app_account_id = Column(
        UUID(as_uuid=True),
        ForeignKey("published_app_accounts.id", ondelete="CASCADE"),
        nullable=True,
        index=True,
    )
    embedded_agent_id = Column(
        UUID(as_uuid=True),
        ForeignKey("agents.id", ondelete="CASCADE"),
        nullable=True,
        index=True,
    )
    external_user_id = Column(String(255), nullable=True, index=True)
    created_by = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), nullable=True, index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    organization = relationship("Organization")
    policy_set = relationship("ResourcePolicySet")
    user = relationship("User", foreign_keys=[user_id])
    published_app_account = relationship("PublishedAppAccount")
    embedded_agent = relationship("Agent")
    creator = relationship("User", foreign_keys=[created_by])

    __table_args__ = (
        Index(
            "uq_resource_policy_assignments_organization_user",
            "organization_id",
            "principal_type",
            "user_id",
            unique=True,
            postgresql_where=and_(
                principal_type == ResourcePolicyPrincipalType.ORGANIZATION_USER,
                user_id != None,
            ),
            sqlite_where=and_(
                principal_type == ResourcePolicyPrincipalType.ORGANIZATION_USER,
                user_id != None,
            ),
        ),
        Index(
            "uq_resource_policy_assignments_app_account",
            "organization_id",
            "principal_type",
            "published_app_account_id",
            unique=True,
            postgresql_where=and_(
                principal_type == ResourcePolicyPrincipalType.PUBLISHED_APP_ACCOUNT,
                published_app_account_id != None,
            ),
            sqlite_where=and_(
                principal_type == ResourcePolicyPrincipalType.PUBLISHED_APP_ACCOUNT,
                published_app_account_id != None,
            ),
        ),
        Index(
            "uq_resource_policy_assignments_embedded_user",
            "organization_id",
            "principal_type",
            "embedded_agent_id",
            "external_user_id",
            unique=True,
            postgresql_where=and_(
                principal_type == ResourcePolicyPrincipalType.EMBEDDED_EXTERNAL_USER,
                embedded_agent_id != None,
                external_user_id != None,
            ),
            sqlite_where=and_(
                principal_type == ResourcePolicyPrincipalType.EMBEDDED_EXTERNAL_USER,
                embedded_agent_id != None,
                external_user_id != None,
            ),
        ),
    )


class ResourcePolicyQuotaCounter(Base):
    __tablename__ = "resource_policy_quota_counters"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    organization_id = Column(UUID(as_uuid=True), ForeignKey("organizations.id", ondelete="CASCADE"), nullable=False, index=True)
    principal_type = Column(
        SQLEnum(ResourcePolicyPrincipalType, values_callable=_enum_values),
        nullable=False,
        index=True,
    )
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=True, index=True)
    published_app_account_id = Column(
        UUID(as_uuid=True),
        ForeignKey("published_app_accounts.id", ondelete="CASCADE"),
        nullable=True,
        index=True,
    )
    embedded_agent_id = Column(
        UUID(as_uuid=True),
        ForeignKey("agents.id", ondelete="CASCADE"),
        nullable=True,
        index=True,
    )
    external_user_id = Column(String(255), nullable=True, index=True)
    model_id = Column(UUID(as_uuid=True), ForeignKey("model_registry.id", ondelete="CASCADE"), nullable=False, index=True)
    quota_window = Column(
        SQLEnum(ResourcePolicyQuotaWindow, values_callable=_enum_values),
        nullable=False,
        default=ResourcePolicyQuotaWindow.MONTHLY,
        server_default=text("'monthly'"),
    )
    period_start = Column(DateTime(timezone=True), nullable=False, index=True)
    period_end = Column(DateTime(timezone=True), nullable=False)
    used_tokens = Column(Integer, nullable=False, default=0, server_default=text("0"))
    reserved_tokens = Column(Integer, nullable=False, default=0, server_default=text("0"))
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now(), nullable=False)

    __table_args__ = (
        UniqueConstraint(
            "principal_type",
            "user_id",
            "published_app_account_id",
            "embedded_agent_id",
            "external_user_id",
            "model_id",
            "quota_window",
            "period_start",
            name="uq_resource_policy_quota_counter_scope_period",
        ),
    )


class ResourcePolicyQuotaReservation(Base):
    __tablename__ = "resource_policy_quota_reservations"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    run_id = Column(UUID(as_uuid=True), nullable=False, unique=True, index=True)
    organization_id = Column(UUID(as_uuid=True), ForeignKey("organizations.id", ondelete="CASCADE"), nullable=False, index=True)
    principal_type = Column(
        SQLEnum(ResourcePolicyPrincipalType, values_callable=_enum_values),
        nullable=False,
        index=True,
    )
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=True, index=True)
    published_app_account_id = Column(
        UUID(as_uuid=True),
        ForeignKey("published_app_accounts.id", ondelete="CASCADE"),
        nullable=True,
        index=True,
    )
    embedded_agent_id = Column(
        UUID(as_uuid=True),
        ForeignKey("agents.id", ondelete="CASCADE"),
        nullable=True,
        index=True,
    )
    external_user_id = Column(String(255), nullable=True, index=True)
    model_id = Column(UUID(as_uuid=True), ForeignKey("model_registry.id", ondelete="CASCADE"), nullable=False, index=True)
    quota_window = Column(
        SQLEnum(ResourcePolicyQuotaWindow, values_callable=_enum_values),
        nullable=False,
        default=ResourcePolicyQuotaWindow.MONTHLY,
        server_default=text("'monthly'"),
    )
    period_start = Column(DateTime(timezone=True), nullable=False, index=True)
    reserved_tokens = Column(Integer, nullable=False, default=0, server_default=text("0"))
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    settled_at = Column(DateTime(timezone=True), nullable=True)

File: frontend-reshet/src/services/resource-policies.ts
import { httpClient } from "./http"

// ============================================================================
// Enums
// ============================================================================

export type ResourcePolicyPrincipalType =
  | "organization_user"
  | "published_app_account"
  | "embedded_external_user"

export type ResourcePolicyResourceType = "agent" | "tool" | "knowledge_store" | "model"

export type ResourcePolicyRuleType = "allow" | "quota"

export type ResourcePolicyQuotaUnit = "tokens"

export type ResourcePolicyQuotaWindow = "monthly"

// ============================================================================
// Response types
// ============================================================================

export interface ResourcePolicyRule {
  id: string
  resource_type: ResourcePolicyResourceType
  resource_id: string
  rule_type: ResourcePolicyRuleType
  quota_unit: ResourcePolicyQuotaUnit | null
  quota_window: ResourcePolicyQuotaWindow | null
  quota_limit: number | null
  created_at: string
  updated_at: string
}

export interface ResourcePolicySet {
  id: string
  name: string
  description: string | null
  is_active: boolean
  created_by: string | null
  created_at: string
  updated_at: string
  included_policy_set_ids: string[]
  rules: ResourcePolicyRule[]
}

export interface ResourcePolicyAssignment {
  id: string
  principal_type: ResourcePolicyPrincipalType
  policy_set_id: string
  user_id: string | null
  published_app_account_id: string | null
  embedded_agent_id: string | null
  external_user_id: string | null
  created_by: string | null
  created_at: string
  updated_at: string
}

// ============================================================================
// Request types
// ============================================================================

export interface CreatePolicySetRequest {
  name: string
  description?: string
  is_active: boolean
}

export interface UpdatePolicySetRequest {
  name?: string
  description?: string
  is_active?: boolean
}

export interface CreatePolicyRuleRequest {
  resource_type: ResourcePolicyResourceType
  resource_id: string
  rule_type: ResourcePolicyRuleType
  quota_unit?: ResourcePolicyQuotaUnit
  quota_window?: ResourcePolicyQuotaWindow
  quota_limit?: number
}

export interface UpdatePolicyRuleRequest {
  resource_id?: string
  quota_unit?: ResourcePolicyQuotaUnit
  quota_window?: ResourcePolicyQuotaWindow
  quota_limit?: number
}

export interface UpsertAssignmentRequest {
  principal_type: ResourcePolicyPrincipalType
  policy_set_id: string
  user_id?: string
  published_app_account_id?: string
  embedded_agent_id?: string
  external_user_id?: string
}

export interface DeleteAssignmentParams {
  principal_type: ResourcePolicyPrincipalType
  user_id?: string
  published_app_account_id?: string
  embedded_agent_id?: string
  external_user_id?: string
}

// ============================================================================
// Service
// ============================================================================

const BASE = "/admin/security/resource-policies"

class ResourcePoliciesService {
  // ── Policy Sets ──

  async listPolicySets(): Promise<ResourcePolicySet[]> {
    return httpClient.get(`${BASE}/sets`)
  }

  async getPolicySet(id: string): Promise<ResourcePolicySet> {
    return httpClient.get(`${BASE}/sets/${id}`)
  }

  async createPolicySet(req: CreatePolicySetRequest): Promise<ResourcePolicySet> {
    return httpClient.post(`${BASE}/sets`, req)
  }

  async updatePolicySet(id: string, req: UpdatePolicySetRequest): Promise<ResourcePolicySet> {
    return httpClient.patch(`${BASE}/sets/${id}`, req)
  }

  async deletePolicySet(id: string): Promise<void> {
    return httpClient.delete(`${BASE}/sets/${id}`)
  }

  // ── Includes ──

  async addInclude(policySetId: string, includedPolicySetId: string): Promise<ResourcePolicySet> {
    return httpClient.post(`${BASE}/sets/${policySetId}/includes`, {
      included_policy_set_id: includedPolicySetId,
    })
  }

  async removeInclude(policySetId: string, includedPolicySetId: string): Promise<void> {
    return httpClient.delete(`${BASE}/sets/${policySetId}/includes/${includedPolicySetId}`)
  }

  // ── Rules ──

  async createRule(policySetId: string, req: CreatePolicyRuleRequest): Promise<ResourcePolicyRule> {
    return httpClient.post(`${BASE}/sets/${policySetId}/rules`, req)
  }

  async updateRule(ruleId: string, req: UpdatePolicyRuleRequest): Promise<ResourcePolicyRule> {
    return httpClient.patch(`${BASE}/rules/${ruleId}`, req)
  }

  async deleteRule(ruleId: string): Promise<void> {
    return httpClient.delete(`${BASE}/rules/${ruleId}`)
  }

  // ── Assignments ──

  async listAssignments(): Promise<ResourcePolicyAssignment[]> {
    return httpClient.get(`${BASE}/assignments`)
  }

  async upsertAssignment(req: UpsertAssignmentRequest): Promise<ResourcePolicyAssignment> {
    return httpClient.put(`${BASE}/assignments`, req)
  }

  async deleteAssignment(params: DeleteAssignmentParams): Promise<void> {
    const qs = new URLSearchParams()
    qs.set("principal_type", params.principal_type)
    if (params.user_id) qs.set("user_id", params.user_id)
    if (params.published_app_account_id) qs.set("published_app_account_id", params.published_app_account_id)
    if (params.embedded_agent_id) qs.set("embedded_agent_id", params.embedded_agent_id)
    if (params.external_user_id) qs.set("external_user_id", params.external_user_id)
    return httpClient.delete(`${BASE}/assignments?${qs.toString()}`)
  }

  // ── Default policy sets ──

  async setPublishedAppDefaultPolicy(publishedAppId: string, policySetId: string | null): Promise<void> {
    return httpClient.patch(`${BASE}/published-apps/${publishedAppId}/default-policy-set`, {
      policy_set_id: policySetId,
    })
  }

  async setEmbeddedAgentDefaultPolicy(agentId: string, policySetId: string | null): Promise<void> {
    return httpClient.patch(`${BASE}/embedded-agents/${agentId}/default-policy-set`, {
      policy_set_id: policySetId,
    })
  }
}

export const resourcePoliciesService = new ResourcePoliciesService()
